### ISSUE TITLE
Keys Generation

### ISSUE THREAD

**Issue by gvenka008c** [spujadas/elk-docker#35]
@spujadas How are you generating these keys? Are these done manually? 

```
#certs/keys for Beats and Lumberjack input
RUN mkdir -p /etc/pki/tls/certs && mkdir /etc/pki/tls/private
ADD ./logstash-forwarder.crt /etc/pki/tls/certs/logstash-forwarder.crt
ADD ./logstash-forwarder.key /etc/pki/tls/private/logstash-forwarder.key
ADD ./logstash-beats.crt /etc/pki/tls/certs/logstash-beats.crt
ADD ./logstash-beats.key /etc/pki/tls/private/logstash-beats.key
```

Thanks,
Govind

**Comment by spujadas** [comment]
Yes the ones in the image were pre-generated and can be found in the GitHub repo.
Everything you want to know about the keys and certificates is here: http://elk-docker.readthedocs.org/#certificates

**Comment by gvenka008c** [comment]
@spujadas Thanks. We are planning to run the elk docker on mesos marathon environment. Just curious what would be the best way to generate the certs? The docker container will be binded to the host that it runs on. We wanted to use hostnames to generate the keys. When the elk container moves from one slave to another slave, not sure how to manage those keys. Thoughts?

**Comment by spujadas** [comment]
Haven't got hands-on experience with Mesos, but I'll give it a shot.
Assuming that your hosts are reachable from your log producers under the names `host1.some.domain`, `host2.some.domain` (i.e. with a common `.some.domain` suffix) etc., the easiest would be to generate certificates with `CN=*.some.domain` as the suffix beforehand, i.e.:

	$ cd /etc/pki/tls
	$ sudo openssl req -x509 -batch -nodes -subj "/CN=*.some.domain/" \
	    -days 3650 -newkey rsa:2048 \
	    -keyout private/logstash-beats.key -out certs/logstash-beats.crt

Then extend the base `sebp/elk` image to add this certificate and private key (see image documentation), and use that image to create the containers.
That way, you can move your containers around and they'll still work even if the hostname changes as long as it remains under `.some.domain`.

Hope that helps, do let me know if I misunderstood the question or if it doesn't work.

**Comment by spujadas** [comment]
Closing following documentation update (https://github.com/spujadas/elk-docker/commit/561e0088b71a68f73246a4141a76107204cc724c) recording the above.
Please feel free to reopen if it turns out that this doesn't properly address your question.

**Comment by gvenka008c** [comment]
@spujadas Thanks.

**Comment by gvenka008c** [comment]
@spujadas Is it a good idea to generate the logstash-beats.key  and logstash-beats.crt inside the docker? I want these to be generate dynamically instead of pre generating it. What would be the best method? Thoughts?

**Comment by spujadas** [comment]
Standard practice (e.g. if you were using a real certificate authority, or CA, rather than self-generating the certificates) says that you should pre-generate the keys and certificates, in order to control how they're being generated and have a limited number of certificates.
(Also, generating a key pair takes some time, so might want to avoid adding that time to your container start-up time. Having said that, on a decent server, for a 2,048- or 4,096-bit RSA key it should take a handful of seconds, so not adding that much overhead on the top of starting the whole ELK stack.)

For instance, with a real CA:
- You wouldn't generally be able to dynamically generate the certificates (among other things because that would be hard to secure).
- Doing so would make your costs explode (cost is usually per certificate).
- If you were generating certificates on-the-fly, then you'd need to revoke them (how being another matter altogether) once the instance is terminated, which involves publishing their serial number in what is known as a certificate revocation list (or CRL). The CRL should be downloaded and checked by log producers, meaning that they would need to download an unnaturally expanding file, which is bad for performance.

With that disclaimer in mind, if you do want to generate the certificates dynamically using a home-made CA, you can, but you would need to proceed differently than what I wrote above. There are plenty of tutorials on OpenSSL that will provide guidance on the exact commands to use, so I'll just give you the general workflow:
- First, you need to pre-generate a self-signed certificate for what is known as a root CA. You add both the key and certificate to the image, and configure all log producers to trust the certificate of this CA (this is the key to the whole thing working).
- Then, you alter the `start.sh` script in the image to:
  - generate a key pair,
  - generate a certificate signature request (or CSR) which can contain the actual hostname of the host or a wildcard `*.your.domain` as the common name (CN),
  - have the CA's certificate produce a certificate based on that CSR.

As you'll note, you still have a hard-coded certificate and key in the image, so if that is what you wanted to avoid in the first place, then this method doesn't work (and in that case you'd need to create an external CA service – on a pair of nodes for high availability – that automatically signs CSRs that it receives, and have the `start.sh` send the CSR to this service). On the other hand, this method does enable you to have certificates containing the actual hostname of the node (e.g. if you don't want to use wildcard certificates), at least if you can work out the hostnames dynamically at runtime.

If "somewhat secure" authentication is not a requirement (e.g. if your network set-up makes sure that only authorised log producers can send logs to ELK) and you want to avoid all this trouble playing around with certificates, then you could also disable the use SSL/TLS (see documentation), and that would definitely work too.

**tl;dr** – It would be very unorthodox, but yes, you could do what you suggested.
